### PR TITLE
feat: dry run AT API transfers

### PR DESCRIPTION
**Pull request by noahjoeris**
This PR
- enables making dry runs on AT API transfers
- adds a generic ActionBanner component. Used for ERC20 spend approvals and the Dry Run
- adds hook useDryRunValidation
- refactors many things

### CODE REVIEW

File: app/src/hooks/useDryRunValidation.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        console.log('Dry Run Successful')
+        addNotification({
+          message: 'Validation Successful. Continue with the transfer.',
+          severity: NotificationSeverity.Success,
+        })
+        setState('Success')
+      } else {
+        console.log('Dry Run Not Successful')
+        addNotification({
+          message: 'Sorry the validation failed. Proceed at your own risk.',
```
Comment (NunoAlexandre) [inline app/src/hooks/useDryRunValidation.tsx:97]:
```suggestion
          message: 'Something is off. Proceed at your own risk.',
```

Comment (NunoAlexandre) [reply app/src/hooks/useDryRunValidation.tsx:97]: A change as a reminder for us to have a more specific language for Turtle. I wonder if we should leave a tip to get in touch with us if the validation failed (like, "Proceed at your own risk or get in touch") but that would be too long.

We should log this into Sentry somehow tho. For instance, if we see multiple dry-runs failing trying to transfer a specific token and/or for a specific source and dest chains, we can work on it right away.

File: app/src/hooks/useDryRunValidation.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      })
+      return false
+    }
+  }, [sourceChain, token, recipient, amount, destinationChain, sender, addNotification])
+
+  const updateHasDryRun = useCallback(async () => {
+    if (!sourceChain || !destinationChain) setHasDryRun(false)
+    else {
+      const route = getRoute(environment, sourceChain, destinationChain)
+      setHasDryRun(route?.sdk === 'AssetTransferApi')
```
Comment (NunoAlexandre) [inline app/src/hooks/useDryRunValidation.tsx:118]: This is easier to read and follow IMO 🙏 

```suggestion
    setHasDryRun(!!sourceChain && !!destinationChain && getRoute(environment, sourceChain, destinationChain)?.sdk === ''AssetTransferApi')
```

File: app/src/utils/routes.ts [hunk 3]
```diff
@@ -102,3 +102,8 @@ export const isTokenAvailableForSourceChain = (
   if (!sourceChain || !token) return false
   return getAllowedTokens(env, sourceChain, null).some(t => t.allowed && t.id === token.id)
 }
+
+export const getRoute = (env: Environment, from: Chain, to: Chain) => {
+  const allRoutes = REGISTRY[env].routes
+  return allRoutes.find(route => route.from === from.uid && route.to === to.uid)
+}
```
Comment (NunoAlexandre) [inline app/src/utils/routes.ts:109]: I implemented this function already in `registry.ts`:

``` typescript
export function getRoute(env: Environment, from: Chain, to: Chain): Route | undefined {
  return REGISTRY[env].routes.find(r => r.from === from.uid && r.to === to.uid)
}
```

IMO there's no extra value in first instantiating `allRoutes`

File: app/src/hooks/useDryRunValidation.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      const result = await atApi.api
+        .tx(txResult.tx)
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        .dryRun(account.address, { signer: account.signer as any })
+
+      if (result.isOk) {
+        console.log('Dry Run Successful')
+        addNotification({
+          message: 'Validation Successful. Continue with the transfer.',
+          severity: NotificationSeverity.Success,
```
Comment (NunoAlexandre) [inline app/src/hooks/useDryRunValidation.tsx:91]:
```suggestion
          message: 'Looking good! Ready when you are,
          severity: NotificationSeverity.Success,
```

wdyt?

**Review by NunoAlexandre — APPROVED** [review]
beautiful work 👌  I really like the refactor to dry the generic `ActionBanner` component and the hook for the dry-run is nicely done too!

Some comments below

**Review by NunoAlexandre — CHANGES_REQUESTED** [review]
didn't mean to approve in the previous review, sorry 🐒

File: app/src/components/ActionBanner.tsx [hunk 5]
```diff
@@ -1,38 +1,45 @@
 import { cn } from '@/utils/cn'
-import Image from 'next/image'
-import { FC } from 'react'
+import { FC, ReactNode } from 'react'
 
-interface Props {
+interface ActionBannerProps {
```
[the remark this answers is not in the document]
Comment (noahjoeris) [reply app/src/components/ActionBanner.tsx:4]: It's convention. But I can reverse it, I personally don't care.
When you export such interfaces you always need to rename it which is annoying. That could be the major reason.

File: app/src/components/Transfer.tsx [hunk 6]
```diff
@@ -87,6 +89,26 @@ const Transfer: FC = () => {
...
+    state: dryRunState,
+    hasDryRun,
+  } = useDryRunValidation({
+    environment,
+    sender: sourceWallet?.sender,
+    sourceChain,
+    token: tokenAmount?.token,
+    recipient: manualRecipient.enabled
+      ? manualRecipient.address
+      : destinationWallet?.sender?.address,
```
[the remark this answers is not in the document]
Comment (noahjoeris) [reply app/src/components/Transfer.tsx]: created a helper

**Comment by noahjoeris** [comment]
I think I addressed everything @NunoAlexandre

File: app/src/components/ActionBanner.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (NunoAlexandre) [reply app/src/components/ActionBanner.tsx]: I see. I don't think we make use of these exported interfaces but I get it. Let's follow the convention then 👍  Thanks Noah

**Review by NunoAlexandre — APPROVED** [review]
Beautiful ✨

**Pull request merged**